adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -683,42 +683,42 @@ private Span newSpan(final String operationName, final Device authenticatedDevic
                     addMicrometerSample(commandContext, timer);
                     Tags.COMPONENT.set(commandContext.getCurrentSpan(), getTypeName());
                     final Command command = commandContext.getCommand();
-
-                    if (!command.isValid()) {
-                        final Exception ex = new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST, "malformed command message");
-                        commandContext.reject(getErrorCondition(ex), 1);
-                        metrics.reportCommand(
-                                command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
-                                sourceAddress.getTenantId(),
-                                ProcessingOutcome.UNPROCESSABLE,
-                                command.getPayloadSize(),
-                                timer);
-                    } else if (!sender.isOpen()) {
-                        commandContext.release(1);
+                    final Future<TenantObject> tenantTracker = getTenantConfiguration(sourceAddress.getTenantId(),
+                            commandContext.getTracingContext());
+
+                    tenantTracker.compose(tenantObject -> {
+                        if (!command.isValid()) {
+                            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                                    "malformed command message"));
+                        } else if (!sender.isOpen()) {
+                            commandContext.release(1);
+                            metrics.reportCommand(
+                                    command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                                    sourceAddress.getTenantId(),
+                                    tenantObject,
+                                    ProcessingOutcome.UNDELIVERABLE,
+                                    command.getPayloadSize(),
+                                    timer);
+                            return Future.succeededFuture();
+                        } else {
+                            return checkMessageLimit(tenantObject, command.getPayloadSize())
+                                    .compose(ok -> {
+                                        onCommandReceived(tenantObject, sender, commandContext);
+                                        return Future.succeededFuture();
+                                    });
+                        }
+                    }).otherwise(failure -> {
+                        // issue credit so that application(s) can send the next command
+                        commandContext.reject(getErrorCondition(failure), 1);
```

Comment:
IMHO this is not yet correct. We reject here based on the assumption that we are dealing with a client induced error, e.g. a malformed command message. However, this code will also be executed if we cannot retrieve the tenant object due to lack of credit (similar to `!sender.isOpen()`). In that case, however, I believe that we would also need to release the command message ...

Reply:
@sophokles73 The Command message is now being _rejected_ if the error is `ClientErrorException` else _released_. I have pushed the changes for _AMQP_ and _MQTT_ adapters.